File: minimumNumberOfPeopleToTeach.go
func minimumTeachings(n int, languages [][]int, friendships [][]int) int {
    m := len(languages)
    
    // Convert user languages to a slice of maps for quick lookup
    userLangs := make([]map[int]bool, m)
    for i := 0; i < m; i++ {
        userLangs[i] = make(map[int]bool)
        for _, lang := range languages[i] {
            userLangs[i][lang] = true
        }
    }

    // Find all friendships where communication is not possible
    toFix := [][]int{}
    for _, pair := range friendships {
        u, v := pair[0]-1, pair[1]-1
        canCommunicate := false
        for lang := range userLangs[u] {
            if userLangs[v][lang] {
                canCommunicate = true
                break
            }
        }
        if !canCommunicate {
            toFix = append(toFix, []int{u, v})
        }
    }

    if len(toFix) == 0 {
        return 0
    }

    minTeach := m // max possible is teaching everyone
    for lang := 1; lang <= n; lang++ {
        teachSet := make(map[int]bool)
        for _, pair := range toFix {
            u, v := pair[0], pair[1]
            if !userLangs[u][lang] {
                teachSet[u] = true
            }
            if !userLangs[v][lang] {
                teachSet[v] = true
            }
        }
        if len(teachSet) < minTeach {
            minTeach = len(teachSet)
        }
    }

    return minTeach
}

File: minimumNumberOfPeopleToTeach.py
from typing import List
from collections import defaultdict

class Solution:
    def minimumTeachings(self, n: int, languages: List[List[int]], friendships: List[List[int]]) -> int:
        m = len(languages)
        # Convert user language lists to sets for fast lookup
        user_languages = [set(langs) for langs in languages]
        
        # Identify all friend pairs that cannot communicate
        to_fix = []
        for u, v in friendships:
            u -= 1
            v -= 1
            if user_languages[u].intersection(user_languages[v]):
                continue
            to_fix.append((u, v))
        
        # For each language, count how many unique users among those in to_fix need to learn it
        min_teach = float('inf')
        for lang in range(1, n + 1):
            teach_set = set()
            for u, v in to_fix:
                if lang not in user_languages[u]:
                    teach_set.add(u)
                if lang not in user_languages[v]:
                    teach_set.add(v)
            min_teach = min(min_teach, len(teach_set))
        
        return 0 if not to_fix else min_teach
